lib/awful/rules.lua

```diff
@@ -188,23 +190,164 @@ function rules.matches_list(c, _rules)
+--
+-- It is important to note that properties themselves have their own
+-- dependencies. For example, a `tag` property implies a `screen`. Therefor, if
+-- a `screen` is already specified, then it will be ignored when the rule is
+-- executed. Properties also have their own priorities. For example, the
+-- `titlebar` and `border_width` need to be applied before the `x` and `y`
+-- positions are set. Otherwise, it will be off or the client will shift
+-- upward everytime Awesome is restarted. A rule source *cannot* change this.
+-- It is up to the callback to be aware of the dependencies and avoid to
+-- introduce issues. For example, if the source want to set a `screen`, it has
+-- to check if the `tag`, `tags` or `new_tag` are on that `screen` or remove
+-- those properties. Otherwise, they will be ignored once the rule is applied.
+--
+-- @tparam string name The provider name. It must be unique.
+-- @tparam function callback The callback
+-- @tparam client callback.c The client
+-- @tparam table callback.properties The current properties. It has to modify
+--  that table.
+-- @tparam table callback.callbacks A table of all callbacks scheduled to be
+--  executed before the main properties are applied.
+-- @tparam[opt={}] table|string dependencies The list of sources this source depends on
+-- @tparam[opt={}] table|string prioritize_over The list of sources this source have a
+--  priority over.
+-- @treturn boolean The position in the list, `nil` if none was found.
+function rules.add_rule_source(name, callback, dependencies, prioritize_over)
+    local min_index, max_index = 1, #rules.rule_sources+1
+
+    dependencies = type(dependencies) == "string"
+        and {dependencies} or dependencies or {}
+
+    prioritize_over = type(prioritize_over) == "string"
+        and {prioritize_over} or prioritize_over or {}
+
+    for k, v in ipairs(rules.rule_sources) do
+        -- Names must be unique
+        assert(v.name ~= name)
+
+        for _, v2 in ipairs(dependencies) do
+            if v.name == v2 then
+                min_index = math.max(min_index, k+1)
             end
         end
+    end
+
+    for k, v in ipairs(rules.rule_sources) do
+        for _, v2 in ipairs(prioritize_over) do
+            if v.name == v2 then
+                max_index = math.min(max_index, k)
+            end
+        end
+    end
+
+    if max_index < min_index then return false end
+
+    -- max_index wins, this give the source the highest possible priority
+    table.insert(rules.rule_sources, max_index, {
+        callback = callback,
+        name     = name
+    })
+
+    return true
+end
+
+--- Remove a source.
+-- @tparam string name The source name.
+-- @treturn boolean If the source was removed
+function rules.remove_rule_source(name)
+    for k, v in ipairs(rules.rule_sources) do
+        if v.name == name then
+            table.remove(rules.rule_sources, k)
+            return true
+        end
+    end
+
+    return false
+end
```

Comment:
I would argue it's bad design

1. It exposes internal structures [1]
2. One of the undisputed advantage of encapsulation is to increase the flexibility of internal structures by exposing a stable interface
3. It moves what's clearly boilerplate code into modules for no good reasons
4. It's only a few lines of code
5. It would force the `foo` in your example to be encoded in an undefined table format (that shall remains undefined)

[1] I really should make it fully ptivate, but I left it undocumented, so it is still private